diff --git a/storage/innobase/xtrabackup/src/backup_mysql.cc b/storage/innobase/xtrabackup/src/backup_mysql.cc
index bc27e0df..328c2086 100644
--- a/storage/innobase/xtrabackup/src/backup_mysql.cc
+++ b/storage/innobase/xtrabackup/src/backup_mysql.cc
@@ -1533,6 +1533,8 @@ bool write_current_binlog_file(MYSQL *connection) {
 
   snprintf(filepath, sizeof(filepath), "%s%c%s", log_bin_dir, FN_LIBCHAR,
            log_status.filename.c_str());
+
+  xtrabackup::utils::prepend_snap_path(filepath);
   result = copy_file(
       ds_data, filepath, log_status.filename.c_str(), 0, FILE_PURPOSE_BINLOG,
       log_status.position + binlog_encryption_header_size(filepath));
@@ -1555,6 +1557,8 @@ bool write_current_binlog_file(MYSQL *connection) {
     ++log_bin_index_filename;
   }
 
+  xtrabackup::utils::prepend_snap_path(log_bin_index);
+
   f_index = fopen(log_bin_index, "r");
   if (f_index == nullptr) {
     xb::error() << "cannot open binlog index file " << SQUOTE(log_bin_index);
diff --git a/storage/innobase/xtrabackup/src/utils.cc b/storage/innobase/xtrabackup/src/utils.cc
index 527d17d1..c5b33ca5 100644
--- a/storage/innobase/xtrabackup/src/utils.cc
+++ b/storage/innobase/xtrabackup/src/utils.cc
@@ -158,5 +158,18 @@ std::string generate_uuid() {
   return uuid_ss.str();
 }
 
+void prepend_snap_path(char *path) {
+  /* if running inside snap, preprend hostfs path */
+  if (getenv("SNAP") != NULL && strncmp(path, "/var/lib/mysql", 14) == 0) {
+    char snap_prefix[FN_REFLEN] = "/var/lib/snapd/hostfs";
+    char *new_path = (char *)malloc(strlen(path) + strlen(snap_prefix) + 2);
+    strcpy(new_path, snap_prefix);
+    strcat(new_path, path);
+    strcpy(path, new_path);
+    free(new_path);
+  }
+}
+
+
 }  // namespace utils
 }  // namespace xtrabackup
diff --git a/storage/innobase/xtrabackup/src/utils.h b/storage/innobase/xtrabackup/src/utils.h
index e8a458dc..4c5d4e50 100644
--- a/storage/innobase/xtrabackup/src/utils.h
+++ b/storage/innobase/xtrabackup/src/utils.h
@@ -47,6 +47,8 @@ unsigned long get_version_number(std::string version_str);
 unsigned long host_total_memory();
 unsigned long host_free_memory();
 
+void prepend_snap_path(char *path);
+
 /** Generates uuid
 @return uuid string */
 std::string generate_uuid();
diff --git a/storage/innobase/xtrabackup/src/xtrabackup.cc b/storage/innobase/xtrabackup/src/xtrabackup.cc
index 2dd5e339..7572ec9b 100644
--- a/storage/innobase/xtrabackup/src/xtrabackup.cc
+++ b/storage/innobase/xtrabackup/src/xtrabackup.cc
@@ -4219,6 +4219,7 @@ void xtrabackup_backup_func(void) {
 #endif
 
   /* cd to datadir */
+  xtrabackup::utils::prepend_snap_path(mysql_real_data_home);
 
   if (my_setwd(mysql_real_data_home, MYF(MY_WME))) {
     xb::error() << "cannot my_setwd " << mysql_real_data_home;
@@ -4878,6 +4879,7 @@ static void stat_with_rec(dict_table_t *table, THD *thd,
 
 static void xtrabackup_stats_func(int argc, char **argv) {
   /* cd to datadir */
+  xtrabackup::utils::prepend_snap_path(mysql_real_data_home);

